bson/src/main/org/bson/codecs/configuration/ProvidersCodecRegistry.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (c) 2008-2014 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.bson.codecs.configuration;
+
+import org.bson.codecs.Codec;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static java.lang.String.format;
+import static org.bson.assertions.Assertions.isTrueArgument;
+
+final class ProvidersCodecRegistry implements CodecRegistry {
+    private final List<CodecProvider> codecProviders;
+
+    ProvidersCodecRegistry(final List<? extends CodecProvider> codecProviders) {
+        isTrueArgument("codecProviders must not be null or empty", codecProviders != null && codecProviders.size() > 0);
+        this.codecProviders = new ArrayList<CodecProvider>(codecProviders);
+    }
+
+    @Override
+    public <T> Codec<T> get(final Class<T> clazz) {
+        return get(new ChildCodecRegistry<T>(this, clazz));
+    }
+
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    <T> Codec<T> get(final ChildCodecRegistry context) {
```

Comment:
Even though we no longer have a withCodec method, we still need the codecs member that was in RootCodecRegistry here.  The reason being that it's the responsibility of the registry to cache Codec instances, not the responsibility of the Provider, which is free to new up a Codec on every request.  And the codecs cache means we don't have to iterate through all the providers on every request.